reactor-netty-http/src/main/java/reactor/netty/http/Cookies.java

```diff
@@ -104,27 +111,35 @@ private Cookies(HttpHeaders nettyHeaders, CharSequence cookiesHeaderName, boolea
 				if (c == null) {
 					continue;
 				}
-				Set<Cookie> existingCookiesOfName = cookies.get(c.name());
-				if (null == existingCookiesOfName) {
-					existingCookiesOfName = new HashSet<>();
-					cookies.put(c.name(), existingCookiesOfName);
-				}
+				Set<Cookie> existingCookiesOfName = cookies.computeIfAbsent(c.name(), k -> new HashSet<>());
 				existingCookiesOfName.add(c);
 			}
 			else {
 				decode = ((ServerCookieDecoder) decoder).decode(aCookieHeader);
 				for (Cookie cookie : decode) {
-					Set<Cookie> existingCookiesOfName = cookies.get(cookie.name());
-					if (null == existingCookiesOfName) {
-						existingCookiesOfName = new HashSet<>();
-						cookies.put(cookie.name(), existingCookiesOfName);
-					}
+					Set<Cookie> existingCookiesOfName = cookies.computeIfAbsent(cookie.name(), k -> new HashSet<>());
 					existingCookiesOfName.add(cookie);
 				}
 			}
 		}
 		cachedCookies = Collections.unmodifiableMap(cookies);
-		state = READ;
+		markReadCookies();
 		return cachedCookies;
 	}
+
+	protected List<String> allCookieHeaders() {
+		return nettyHeaders.getAll(cookiesHeaderName);
+	}
+
+	protected final boolean hasReadCookies() {
+		return state == READ;
```

Comment:
the atomic field updater can only guarantee atomicity and memory-ordering of the write operations if all write operations are performed through it rather than the volatile variable.

for that reason, this should be `STATE.set(this, READ)`.

(reading from the volatile is always ok)

Reply:
but here `hasReadCookies` we want just to check the state we do not want to set it or I miss something?

Reply:
oh gods, that's right 🤦 I was looking on the left hand side seeing `state = READ` and here `state == READ` and my brain jumped to conclusions 😆

all good, nothing to see here, move along

Reply:
thanks a lot